src/TSnePlotView.js

```diff
@@ -16,97 +16,108 @@ class ExperimentPageView extends React.Component {
+        .then((responseJson) => {
+          this.setState({
+            geneExpressionData: responseJson,
+            geneExpressionErrorMessage: null,
+            loadingGeneExpression: false,
+          })
         })
-      })
-      .catch((reason) => {
-        this.setState({
-          errorMessage: `${reason.name}: ${reason.message}`,
-          loadingClusters: false,
-          loadingGeneExpression: false
+        .catch((reason) => {
+          this.setState({
+            geneExpressionErrorMessage: `${reason.name}: ${reason.message}`,
+            loadingGeneExpression: false
+          })
         })
-      })
+    })
   }
 
   componentWillReceiveProps(nextProps) {
-    if (nextProps.atlasUrl !== this.props.atlasUrl ||  // First two will never happen but it’s the right thing to do
-        nextProps.experimentAccession !== this.props.experimentAccession ||
-        nextProps.perplexity !== this.props.perplexity ||
-        nextProps.k !== this.props.k) {
-
-      this.setState({
-        loadingClusters: true,
-        loadingGeneExpression: true
-      })
-      this._fetchAndSetState(nextProps)
-
+    if (nextProps.perplexity !== this.props.perplexity) {
+      this._fetchAndSetStateCellClusters(nextProps)
+      this._fetchAndSetStateGeneId(nextProps)
+    } else if (nextProps.k !== this.props.k) {
+      this._fetchAndSetStateCellClusters(nextProps)
     } else if (nextProps.geneId !== this.props.geneId) {
-
-      this.setState({
-        loadingGeneExpression: true
-      })
-      this._fetchAndSetState(nextProps)
-
+      this._fetchAndSetStateGeneId(nextProps)
     }
   }
 
   componentDidMount() {
-    this.setState({
-      loadingClusters: true,
-      loadingGeneExpression: true
-    })
-    // Having _fetchAndSetState as callback is the right thing, but then we can’t return the promise; see tests
-    return this._fetchAndSetState(this.props)
+    this._fetchAndSetStateCellClusters(this.props)
+    this._fetchAndSetStateGeneId(this.props)
   }
 
   render() {
     const {height, atlasUrl, resourcesUrl} = this.props
     const {suggesterEndpoint, geneId, speciesName, highlightClusters, ks, k, perplexities, perplexity} = this.props
     const {onChangePerplexity, onChangeK, onSelectGeneId} = this.props
-    const {loadingClusters, loadingGeneExpression, data, errorMessage} = this.state
+    const {loadingGeneExpression, geneExpressionData, geneExpressionErrorMessage} = this.state
+    const {loadingCellClusters, cellClustersData, cellClustersErrorMessage} = this.state
 
     return (
       <div className={`row`}>
         <div className={`small-12 medium-6 columns`}>
           <ClusterTSnePlot height={height}
-                           plotData={data}
+                           plotData={cellClustersData}
                            perplexities={perplexities}
                            perplexity={perplexity}
```

Comment:
Would it be better if `perplexity` was renamed to something like `selectedPerplexity`? The same could be said about `k`.